block/manager.go

```diff
@@ -688,18 +678,64 @@ func (m *Manager) applyBlock(ctx context.Context, header *types.SignedHeader, da
 	return m.execApplyBlock(ctx, m.lastState, header, data)
 }
 
-func (m *Manager) execValidate(_ types.State, _ *types.SignedHeader, _ *types.Data) error {
-	// TODO(tzdybal): implement
-	return nil
+func (m *Manager) Validate(ctx context.Context, header *types.SignedHeader, data *types.Data) error {
+	m.lastStateMtx.RLock()
+	defer m.lastStateMtx.RUnlock()
+	return m.execValidate(m.lastState, header, data)
 }
 
-func (m *Manager) execCommit(ctx context.Context, newState types.State, h *types.SignedHeader, _ *types.Data) ([]byte, error) {
-	err := m.exec.SetFinal(ctx, h.Height())
-	return newState.AppHash, err
+// execValidate validates a pair of header and data against the last state
+func (m *Manager) execValidate(lastState types.State, header *types.SignedHeader, data *types.Data) error {
+	// Validate the basic structure of the header
+	if err := header.ValidateBasic(); err != nil {
+		return fmt.Errorf("invalid header: %w", err)
+	}
+
+	// Validate the header against the data
+	if err := types.Validate(header, data); err != nil {
+		return fmt.Errorf("validation failed: %w", err)
+	}
+
+	// Ensure the header's Chain ID matches the expected state
+	if header.ChainID() != lastState.ChainID {
+		return fmt.Errorf("chain ID mismatch: expected %s, got %s", lastState.ChainID, header.ChainID())
+	}
+
+	// Check that the header's height is the expected next height
+	expectedHeight := lastState.LastBlockHeight + 1
+	if header.Height() != expectedHeight {
+		return fmt.Errorf("invalid height: expected %d, got %d", expectedHeight, header.Height())
+	}
+
+	// Verify that the header's timestamp is strictly greater than the last block's time
+	headerTime := header.Time()
+	if headerTime.Before(lastState.LastBlockTime) || headerTime.Equal(lastState.LastBlockTime) {
+		return fmt.Errorf("block time must be strictly increasing: got %v, last block time was %v",
+			headerTime, lastState.LastBlockTime)
+	}
+
+	// Validate that the header's AppHash matches the lastState's AppHash
+	// Note: Assumes deferred execution
+	if !bytes.Equal(header.AppHash, lastState.AppHash) {
+		return fmt.Errorf("app hash mismatch: expected %x, got %x", lastState.AppHash, header.AppHash)
+	}
```

Comment:
this check we should leave up to execution, we shouldnt make assumptions on execution in rollkit.

Reply:
how do we do that though? Like the call to `ExecuteTxs` checks the app hash before application? The execution layer would then need to know the appHash to validate against right?